### ISSUE TITLE
[Bug]: ESLint `lintDirtyModulesOnly` not working properly, still lints everything

### ISSUE THREAD

**Issue by markmssd** [web-infra-dev/rspack#7748]
### System Info

  System:
    OS: macOS 14.6.1
    CPU: (10) arm64 Apple M1 Max
    Memory: 631.08 MB / 32.00 GB
    Shell: 5.9 - /bin/zsh
  Binaries:
    Node: 20.14.0 - ~/.nvm/versions/node/v20.14.0/bin/node
    Yarn: 1.22.22 - ~/.nvm/versions/node/v20.14.0/bin/yarn
    npm: 10.8.2 - ~/.nvm/versions/node/v20.14.0/bin/npm
    pnpm: 7.28.0 - ~/Library/pnpm/pnpm
    bun: 1.1.8 - ~/.bun/bin/bun
  Browsers:
    Brave Browser: 126.1.67.123
    Chrome: 128.0.6613.113
    Edge: 128.0.2739.54
    Safari: 17.6
  npmPackages:
    @rspack/cli: 1.0.0 => 1.0.0 
    @rspack/core: 1.0.0 => 1.0.0 

### Details

Might be related to https://github.com/web-infra-dev/rspack/issues/6432

Even if `lintDirtyModulesOnly` is enabled, it runs ESLint on all files. The expected behaviour (which seems to work well with Webpack) is to run the linter on the modified file only.

It's a negligible difference for a small repo like `rspack-repro`, but in large repos, it causes HMR to reload in multiple seconds, instead of <1sec.

### Currently, modifying a single file will run the linter on all of them:

<img width="888" alt="Screenshot 2024-08-31 at 4 03 33 PM" src="https://github.com/user-attachments/assets/d08dc0ed-1824-469f-ba7d-1e8f923ab9e8">

### Expected (when using [eslint-webpack-plugin](https://github.com/markmssd/rspack-repro/blob/eslint-warnings/config.mjs#L5)):

<img width="865" alt="Screenshot 2024-08-31 at 4 04 23 PM" src="https://github.com/user-attachments/assets/981f64d3-6f0f-4ea9-b997-cab15284ad4c">

### Reproduce link

https://github.com/markmssd/rspack-repro/tree/eslint-warnings

### Reproduce Steps

```
git clone https://github.com/markmssd/rspack-repro/tree/eslint-warnings
cd eslint-warnings
npm install
npm run dev:rspack
```

Now, modify `src/index.js`. It will run ESLint on all files (`src/index.js` and `src/render.js`), even if `lintDirtyModulesOnly` is enabled in `config.mjs`.

**Comment by markmssd** [comment]
I digged more into it, it happens because it now uses the [finishModules](https://webpack.js.org/api/compilation-hooks/#finishmodules) hook instead of the [succeedModule](https://webpack.js.org/api/compilation-hooks/#succeedmodule) (which is done on purpose for performance reasons it seems?).

The problem when I use `succeedModule` instead is that the linter always receives the original file content, not the updated one, so even if an issue is fixed, it will still report the issue. Maybe an issue from Rspack?

**Comment by LingyuCoder** [comment]
Fixed by @rspack/core@1.0.8 and eslint-rspack-plugin@4.2.1
